### PR TITLE
smt: Switch to SHA-256 hashing

### PR DESCRIPTION
**Pull request by popcnt1**
## Summary

Using double sha256(what bitcoin merkel tree has) replacing sha3 in smt:

1. as safe as bitcoin
2. much faster

I've tested using sha256(double sha256 still much faster) in smt:

~20% faster in statedb/genesis process(351s vs 426s):

10M utxo + 3M inscirption

sha256 is 4x-5x faster:

```
test rust_crypto_sha3_256_input_32_bytes   ... bench:         234.56 ns/iter (+/- 4.92) = 136 MB/s
test rust_crypto_sha3_256_input_4096_bytes ... bench:       6,951.77 ns/iter (+/- 132.45) = 589 MB/s
test sha256_32                             ... bench:          43.64 ns/iter (+/- 0.08) = 744 MB/s
test sha256_4096                           ... bench:       1,713.34 ns/iter (+/- 2.77) = 2391 MB/s
test tiny_keccak_sha3_256_input_32_bytes   ... bench:         236.14 ns/iter (+/- 6.98) = 135 MB/s
test tiny_keccak_sha3_256_input_4096_bytes ... bench:       6,965.95 ns/iter (+/- 125.54) = 588 MB/s
```

Is there any particular reason why we should use keccak? @jolestar @baichuan3 

sha256 is enough strong and widely used, bitcoin use it too.

-------

I've confirmed sha256 is well implemented with SHA exetension inscturtion set in rust. It could be make faster with AVX512 but this is not so widely deployed.

### CODE REVIEW

**Comment by jolestar** [comment]
There is no special reason, but this is a break change. I recorded it in #2402, and we can discuss it in the dev meeting.

**Comment by feliciss** [comment]
> Is there any particular reason why we should use keccak? @jolestar @baichuan3

I remember there's a particular reason that Rooch uses it for Ethereum at first.

> sha256 is enough strong and widely used, bitcoin use it too.

Later, has rooch switched to Bitcoin Layer 2 solutions?

**Comment by popcnt1** [comment]
&gt; &gt; Is there any particular reason why we should use keccak? @jolestar @baichuan3
&gt; 
&gt; I remember there's a particular reason that Rooch uses it for Ethereum at first.
&gt; 
&gt; &gt; sha256 is enough strong and widely used, bitcoin use it too.
&gt; 
&gt; Later, has rooch switched to Bitcoin Layer 2 solutions?



> > Is there any particular reason why we should use keccak? @jolestar @baichuan3
> 
> I remember there's a particular reason that Rooch uses it for Ethereum at first.
> 
> > sha256 is enough strong and widely used, bitcoin use it too.
> 
> Later, has rooch switched to Bitcoin Layer 2 solutions?

Yes. What‘s your suggestion? Replace it with sha256?

**Comment by feliciss** [comment]
> > > Is there any particular reason why we should use keccak? @jolestar @baichuan3 > > I remember there's a particular reason that Rooch uses it for Ethereum at first. > > > sha256 is enough strong and widely used, bitcoin use it too. > > Later, has rooch switched to Bitcoin Layer 2 solutions?
> 
> > > Is there any particular reason why we should use keccak? @jolestar @baichuan3
> > 
> > 
> > I remember there's a particular reason that Rooch uses it for Ethereum at first.
> > > sha256 is enough strong and widely used, bitcoin use it too.
> > 
> > 
> > Later, has rooch switched to Bitcoin Layer 2 solutions?
> 
> Yes. What‘s your suggestion? Replace it with sha256?

I second on this.

**Comment by baichuan3** [comment]
smt hash using sha256, address hash remain keccak

File: moveos/smt/src/jellyfish_merkle/hash.rs [hunk 1]
```diff
@@ -10,16 +16,12 @@ use primitive_types::H256;
...
-    fmt::{self, Debug},
-    str::FromStr,
-};
 use tiny_keccak::{Hasher, Sha3};
 
 pub(crate) fn merkle_hash(left: HashValue, right: HashValue) -> HashValue {
     let mut value = left.to_vec();
     value.extend(right.to_vec());
-    HashValue::sha3_256_of(&value)
+    HashValue::sha256_of(&value)
```
Comment (jolestar) [inline moveos/smt/src/jellyfish_merkle/hash.rs:24]: Bitcoin TagedHash

Comment (popcnt1) [reply moveos/smt/src/jellyfish_merkle/hash.rs]: > Bitcoin TagedHash

tagged hash is used for signing, in bitcoin merkel tree is double sha256:

sha256(sha256(data))

File: moveos/smt/src/jellyfish_merkle/hash.rs [hunk 2]
```diff
@@ -75,46 +73,17 @@ impl HashValue {
...
-    {
-        let mut sha3 = Sha3::v256();
-        for buffer in buffers {
-            sha3.update(buffer);
-        }
-        HashValue::from_keccak(sha3)
+    /// Creates a new `HashValue` from a byte array by applying SHA-256 followed by SHA-256.
+    pub fn double_sha256(data: &[u8]) -> Self {
+        let digest = Sha256::digest(Sha256::digest(data).digest);
+        HashValue::new(digest.digest)
```
Comment (jolestar) [inline moveos/smt/src/jellyfish_merkle/hash.rs:79]: Use the SHA256t from bitcoin-hashs? We can add the bitcoin-hashs dependency.

**Review by jolestar — APPROVED** [review]

**Pull request merged**